File: mcp_gateway/src/audit.ts
import { appendFileSync, existsSync, mkdirSync, renameSync, statSync, unlinkSync } from "fs";
import { dirname } from "path";
import { homedir } from "os";

export interface AuditEntry {
  action: string;
  client_id: string;
  reason: string;
  fields: Record<string, string>;
  timestamp?: number;
}

const AUDIT_PATH = process.env.AUDIT_LOG ?? `${homedir()}/.agent-sandbox/audit.jsonl`;
const MAX_SIZE = parseInt(process.env.AUDIT_MAX_SIZE ?? String(10 * 1024 * 1024));

export function log(entry: AuditEntry): void {
  try {
    const dir = dirname(AUDIT_PATH);
    if (!existsSync(dir)) mkdirSync(dir, { recursive: true });

    rotateIfNeeded();

    const record = { ...entry, timestamp: entry.timestamp ?? Date.now() / 1000 };
    appendFileSync(AUDIT_PATH, JSON.stringify(record) + "\n");
  } catch (e) {
    console.error(`Audit write failed: ${e}`);
  }
}

function rotateIfNeeded(): void {
  if (!existsSync(AUDIT_PATH)) return;
  const size = statSync(AUDIT_PATH).size;
  if (size < MAX_SIZE) return;

  const rotated = AUDIT_PATH + ".1";
  if (existsSync(rotated)) unlinkSync(rotated);
  renameSync(AUDIT_PATH, rotated);
}

File: http_proxy/shared/audit.py
"""Structured JSONL audit logging with size-based rotation."""

import json
import logging
import os
import time
from dataclasses import dataclass, asdict
from pathlib import Path

logger = logging.getLogger("audit")

AUDIT_PATH = Path(os.environ.get("AUDIT_LOG", "/var/log/sandbox/audit.jsonl"))
MAX_SIZE_BYTES = int(os.environ.get("AUDIT_MAX_SIZE", str(10 * 1024 * 1024)))  # 10MB


@dataclass
class AuditEntry:
    action: str          # ALLOW, DENY, APPROVED, DENIED
    client_id: str
    reason: str
    fields: dict         # domain/method/path or tool/params
    timestamp: float = 0

    def __post_init__(self):
        if not self.timestamp:
            self.timestamp = time.time()


def log(entry: AuditEntry) -> None:
    """Append entry to audit log. Rotate if over max size."""
    try:
        AUDIT_PATH.parent.mkdir(parents=True, exist_ok=True)
        _rotate_if_needed()
        with open(AUDIT_PATH, "a") as f:
            f.write(json.dumps(asdict(entry)) + "\n")
    except Exception as e:
        logger.error(f"Audit write failed: {e}")


def _rotate_if_needed() -> None:
    """Rename current log to .1 if over size limit."""
    if not AUDIT_PATH.exists():
        return
    if AUDIT_PATH.stat().st_size < MAX_SIZE_BYTES:
        return

    rotated = AUDIT_PATH.with_suffix(".jsonl.1")
    if rotated.exists():
        rotated.unlink()
    AUDIT_PATH.rename(rotated)
